doc/md/fundementals/icp-features/6-system-functions.md

```diff
@@ -0,0 +1,177 @@
+
+- [`heartbeat`](https://internetcomputer.org/docs/references/ic-interface-spec#heartbeat)
+- [`timer`](https://internetcomputer.org/docs/references/ic-interface-spec#global-timer)
+- [`preupgrade`](#preupgrade)
+- [`postupgrade`](#postupgrade)
+- [`inspect`](https://internetcomputer.org/docs/references/ic-interface-spec#system-api-inspect-message)
+
+Declaring any other system function will result in an error. Canisters can use these functions to efficiently manage state transitions, automate tasks, or handle system-level operations. 
+
+## `heartbeat()`
+
+Canisters can opt to receive [heartbeat messages](https://internetcomputer.org/docs/building-apps/network-features/periodic-tasks-timers#heartbeats) by exposing a `canister_heartbeat` function. In Motoko, this is achieved by declaring the system function `heartbeat`, which takes no arguments and returns an asynchronous unit type (`async ()`).
+
+Since `heartbeat()` is async, it can invoke other asynchronous functions and await their results. This function executes on every **subnet heartbeat**, enabling periodic task execution without requiring external triggers. Since subnet heartbeats operate at the protocol level, their timing is not precise and depends on network conditions and execution load. As a result, using heartbeats for high-frequency or time-sensitive operations should be done cautiously, as there is no guarantee of real-time execution.
+
+Every async call in Motoko causes a context switch, which means the actual execution of the heartbeat function may be delayed relative to when the subnet triggers it. The function’s result is ignored, so any errors or traps during execution do not impact future heartbeat calls.
+
+If a canister exports a function named `canister_heartbeat`, it must have the type `() -> ()`, ensuring it adheres to the expected [system function signature](https://internetcomputer.org/docs/references/ic-interface-spec#heartbeat).
+
+One example use case for [heartbeats](https://internetcomputer.org/docs/building-apps/network-features/periodic-tasks-timers#heartbeats) is a recurring alarm, where the canister performs an action at regular intervals. In the example below, the canister triggers an event every fifth heartbeat by maintaining a counter.
+
+```motoko no-repl
+import Debug "mo:base/Debug";
+
+persistent actor Alarm {
+
+  let n = 5;
+  var count = 0;
+
+  public shared func ring() : async () {
+    Debug.print("Ring!");
+  };
+
+  system func heartbeat() : async () {
+    if (count % n == 0) {
+      await ring();
+    };
+    count += 1;
+  }
+}
+```
+
+## `timer()`
+
+The [`timer()` system function](https://internetcomputer.org/docs/building-apps/network-features/periodic-tasks-timers#timers) lets canisters schedule a task to execute after a specified delay. To make the timer repeat, the function must explicitly call `setGlobalTimer()` within its body to reset the timer. It accepts a single argument to set the global timer and returns `async ()`.
```

Comment:
Actually, `system timer()` gets called with the `setGlobalTimer` function as an argument after install or upgrade. Then the canister can choose to remember that (and call it later) or call it immediately to set the next expiration.